lib/csu/aarch64c/reloc.c

```diff
@@ -0,0 +1,104 @@
+/*-
+ * Copyright (c) 2019 Leandro Lupori
+ * Copyright (c) 2021 The FreeBSD Foundation
+ * Copyright (c) 2022 Jessica Clarke
+ *
+ * Portions of this software were developed by Andrew Turner
+ * under sponsorship from the FreeBSD Foundation.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+__FBSDID("$FreeBSD$");
+
+/*
+ * Fragments consist of a 64-bit address followed by a 56-bit length and an
+ * 8-bit permission field.
+ */
+static uintcap_t
+init_cap_from_fragment(const Elf_Addr *fragment, void * __capability data_cap,
```

Comment:
This could be specialised for now to only handle MORELLO_FRAG_EXECUTABLE, but Morello will eventually ditch caprelocs and emit R_MORELLO_RELATIVE for static binaries (I think the MR for that is still open, but it adds it behind a flag so not an ABI break), which will mean we need the full implementation here, so I've just kept it almost an exact copy of rtld's (the only difference is rtld has DATA/FUNC_PTR_REMOVE_PERMS aliases for the permissions bitsets that have been inlined here, as is done in the kernel).